Renesas R-Car: Add Pinmux driver

julien-massot opened the pull request:
This PR allows to select alternate functions for Renesas R-Car multiplexed pins.
With this driver will be able to add CAN and Serial driver.

Each SoC of the gen 3 family will have to define the PFC configuration for a set of pin:
e.g 
can0_data_a, scif1_data_a, scif2_data_a
Then a board will be able to enable these function thanks to pinmux_rcar_set_pingroup.

aaillet reviewed (approved):
This driver is already used in our public fork for a long time.
It has been used on h3ulcb board during many tests and works well.
Thanks Julien !

julien-massot commented:
Hi @mnkp , this pinmux driver is important for the Renesas platform.
Does it's possible that you have a look at it ?

galak commented on soc/arm/renesas_rcar/gen3/pfc-r8a77951.h:
```diff
@@ -0,0 +1,26 @@
+#define _PFC_RCAR_R8A77951_
+#include "pfc-rcar.h"
+
+static const struct rcar_pin can0_data_a[] = {
```
We should encode these in devicetree.  Look at what STM32, NXP Kinetis and Atmel are doing here.

julien-massot replied:
ok will do, thanks for the review @galak  !

galak reviewed (dismissed):
We need to encode the pindata in devicetree instead of in a header.

mnkp commented on drivers/pinmux/pinmux_rcar.c:
```diff
@@ -0,0 +1,62 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT renesas_rcar_pinmux
```
`drivers/pinmux/pinmux_rcar.c` is not an actual driver since it doesn't implement any public API functions. That's OK, since we're going to deprecate the pinmux API. Let's move this file to `soc/arm/renesas_rcar/gen3/`. As such the pinmux functions will be exposed to rcar drivers only. The public pinmux/pinctrl API can be added later.

julien-massot replied:
Thanks, I did move the pinmux driver to the soc directory in the second version of this PR.

Commit:
soc: arm: renesas_rcar: add pinmux driver

Allow to set an alternate function for a given pin configuration,
GPSR register allow to set a pin to GPIO or peripheral, IPSR allows
to choose an alternate function.

Other pin configurations such as pull-up are not supported at the
moment, but could be added later if useful.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

Commit:
dts: bindings: pinctrl: add Renesas R-Car Pinmux

Renesas R-Car series contains a PFC (Pin Function Controller).
This module consists of registers for selecting the function of
the multiplexed pins and controls the pull-up resistor on each pin.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

Commit:
dts: arm: r8a77951: add pin-controller node

Pin controller address is the same for all member of the
GEN3 SoC Series, but pinmux configuration is SoC specific.

This patch defines pinmux configuration for
SCIF(UART) and CAN0 module.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

Commit:
driver: can: renesas: add pinctrl-0 property

Get pin configuration from device tree to
set alternate pin function.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

julien-massot commented:
Hi @galak, @mnkp does it's possible that you have a look again at this patchset ?

carlescufi commented:
> Marking this DNM as we are hopefully going to have a new pinctrl API so want to make sure whatever is done here is in sync with that. We can decide if it makes sense to merge this and have follow up PRs or what.

I think we should merge this given that the new pinctrl API may not be ready in time

mbolivar-nordic reviewed (dismissed):
Please use the new pinctrl helpers documented in https://docs.zephyrproject.org/latest/reference/devicetree/api.html#pinctrl-pin-control

julien-massot commented:
> Please use the new pinctrl helpers documented in https://docs.zephyrproject.org/latest/reference/devicetree/api.html#pinctrl-pin-control

Just changed DT_PHANDLE_BY_IDX by DT_PINCTRL_0. Thanks !

gmarull reviewed (changes requested):
A few questions and change requests. I'd wait a bit until `pinctrl` API is merged, and see if these SoCs fit into the proposed model or some extensions are required.

mbolivar-nordic commented on dts/arm/renesas/gen3/r8a77951.dtsi:
```diff
@@ -5,3 +5,36 @@
 #include <renesas/gen3/rcar_gen3_cr7.dtsi>
+#include <dt-bindings/pinctrl/rcar-pinctrl.h>
+
+&pfc {
```
>> This driver and dts bindings differs from the Linux one
>
> I'm not referring to the implementation, but how things are described in Devicetree.

Regarding this, please see the general upstream binding rules we recently added to the documentation:

https://docs.zephyrproject.org/latest/guides/dts/bindings.html#check-for-existing-bindings

In particular, see this requirement:

> , if there is an existing binding for your device in an authoritative location, you should try to replicate its properties when writing a Zephyr binding, and you must justify any Zephyr-specific divergences.

I'd like to see the divergence justified. I understand the complexity concerns in source code, but can the existing sources be implemented in terms of a subset of the upstream binding? If not, why not?

julien-massot replied:
Hi thanks for your comments.
I implemented pull-up and pull-down features, as well as drive strength settings.
I'm still fare from the Linux upstream binding, I tried to explain the reasons and the difficulties.
Since I did a lot of modifications to the driver and the binding I created a new pull request #40550 which
supersedes this one.
